File: src/PersianTextGuard/SourceMap.cs
namespace PersianTextGuard;

/// <summary>A reading of a message, with where each of its characters came from in the message.</summary>
/// <remarks>
/// Two maps rather than one, so a fallback can widen in both directions: a match starting at
/// reading index <c>i</c> starts no later than <c>StartMap[i]</c>, and one ending at <c>j</c> ends no
/// earlier than <c>EndMap[j]</c>. When the mapping is exact they are the same array.
/// </remarks>
internal sealed class MappedText
{
    public MappedText(string text, int[] startMap, int[] endMap)
    {
        Text = text;
        StartMap = startMap;
        EndMap = endMap;
    }

    /// <summary>The reading, identical to the one the filter searched.</summary>
    public string Text { get; }

    /// <summary>For each reading character, the earliest message index a match starting there may start at.</summary>
    public int[] StartMap { get; }

    /// <summary>For each reading character, the latest message index a match ending there may end at.</summary>
    public int[] EndMap { get; }
}

/// <summary>
/// Rebuilds the filter's lossy readings of a message with source maps, so a position found in a
/// reading can be turned into a position in the message as it was passed.
/// </summary>
/// <remarks>
/// Maps are only built for a message that contained a match, and only for the reading the match
/// came from, so clean messages never pay for them.
/// </remarks>
internal static class SourceMap
{
    /// <summary>The <paramref name="kind"/> reading of <paramref name="original"/>, with its maps.</summary>
    /// <remarks>
    /// The mapped reading is compared with the reading the filter searches. They only differ if
    /// segment-wise normalization disagrees with whole-string normalization, in which case the maps
    /// fall back to whole whitespace-separated chunks and then to the whole message: a coarser map
    /// can hide more than the word, but never less.
    /// </remarks>
    internal static MappedText Build(string original, ReadingKind kind)
    {
        var map = new List<int>(original.Length);
        var text = PersianNormalizer.Normalize(original, PersianNormalization.Comparison, map);
        var expected = PersianNormalizer.Normalize(original);

        if (kind is ReadingKind.Folded or ReadingKind.FoldedSqueezed)
        {
            text = Apply(text, ref map, static (value, stepMap) => ProfanityFilter.Fold(value, stepMap));
            expected = ProfanityFilter.Fold(expected);
        }

        if (kind is ReadingKind.Squeezed or ReadingKind.FoldedSqueezed)
        {
            text = Apply(text, ref map, static (value, stepMap) => ProfanityFilter.Squeeze(value, stepMap));
            expected = ProfanityFilter.Squeeze(expected);
        }

        if (text == expected)
        {
            var exact = map.ToArray();
            return new MappedText(text, exact, exact);
        }

        return ChunkMap(original, expected);
    }

    /// <summary>
    /// Maps each whitespace-separated chunk of <paramref name="reading"/> to the matching chunk of
    /// <paramref name="original"/> as a whole; the whole message when the chunk counts differ.
    /// </summary>
    internal static MappedText ChunkMap(string original, string reading)
    {
        var originalChunks = Chunks(original);
        var readingChunks = Chunks(reading);
        if (originalChunks.Count != readingChunks.Count)
        {
            return WholeMessageMap(original, reading);
        }

        var starts = new int[reading.Length];
        var ends = new int[reading.Length];
        var chunk = 0;
        var start = 0;
        var end = 0;

        for (var i = 0; i < reading.Length; i++)
        {
            if (chunk < readingChunks.Count && i >= readingChunks[chunk].Start)
            {
                start = originalChunks[chunk].Start;
                end = originalChunks[chunk].Last;
                if (i == readingChunks[chunk].Last)
                {
                    chunk++;
                }
            }

            // Separators keep the previous chunk's values, which keeps both maps non-decreasing.
            starts[i] = start;
            ends[i] = end;
        }

        return new MappedText(reading, starts, ends);
    }

    /// <summary>Maps every reading character to the whole message.</summary>
    internal static MappedText WholeMessageMap(string original, string reading)
    {
        var starts = new int[reading.Length];
        var ends = new int[reading.Length];
        var last = Math.Max(0, original.Length - 1);

        for (var i = 0; i < ends.Length; i++)
        {
            ends[i] = last;
        }

        return new MappedText(reading, starts, ends);
    }

    private static string Apply(string text, ref List<int> map, Func<string, List<int>, string> step)
    {
        var stepMap = new List<int>(text.Length);
        var result = step(text, stepMap);

        var composed = new List<int>(stepMap.Count);
        foreach (var index in stepMap)
        {
            composed.Add(map[index]);
        }

        map = composed;
        return result;
    }

    private static List<(int Start, int Last)> Chunks(string text)
    {
        var chunks = new List<(int Start, int Last)>();
        var start = -1;

        for (var i = 0; i <= text.Length; i++)
        {
            var separator = i == text.Length || char.IsWhiteSpace(text[i]);
            if (!separator && start < 0)
            {
                start = i;
            }
            else if (separator && start >= 0)
            {
                chunks.Add((start, i - 1));
                start = -1;
            }
        }

        return chunks;
    }
}

File: python/src/persian_text_guard/_source_map.py
"""Where each character of a reading came from in the message. Port of ``js/src/source-map.ts``.

Which ports ``dotnet/src/PersianTextGuard/SourceMap.cs``. Positions are units of the UTF-16 view.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from enum import IntEnum

from ._fold import fold, squeeze
from ._normalizer import COMPARISON_STEPS, normalize_with_map
from ._unicode import is_white_space


class ReadingKind(IntEnum):
    """The lossy forms of a message the filter searches, in the order it searches them."""

    NORMALIZED = 0
    """Comparison-normalized."""

    SQUEEZED = 1
    """Normalized, with repeated letters squeezed to one."""

    FOLDED = 2
    """Normalized, with look-alike characters folded."""

    FOLDED_SQUEEZED = 3
    """Normalized, folded, then squeezed."""


@dataclass(frozen=True, slots=True)
class MappedText:
    """A reading of a message, with where each of its characters came from in the message.

    Two maps rather than one, so a fallback can widen in both directions: a match starting at reading
    index ``i`` starts no later than ``start_map[i]``, and one ending at ``j`` ends no earlier than
    ``end_map[j]``. When the mapping is exact they are the same list.

    Attributes:
        text: The reading, identical to the one the filter searched.
        start_map: For each reading unit, the earliest message index a match starting there may start at.
        end_map: For each reading unit, the latest message index a match ending there may end at.
    """

    text: str
    start_map: list[int]
    end_map: list[int]


MapCache = list[MappedText | None]

_Step = Callable[[str, list[int]], str]


def _fold_step(value: str, map_: list[int]) -> str:
    return fold(value, map_)


def _squeeze_step(value: str, map_: list[int]) -> str:
    return squeeze(value, map_)


def new_map_cache() -> MapCache:
    """A fresh cache with one slot per :class:`ReadingKind`."""
    return [None, None, None, None]


def build(original: str, kind: ReadingKind, cache: MapCache | None = None) -> MappedText:
    """The ``kind`` reading of ``original``, with its maps, reusing and filling ``cache``.

    Each reading is mapped once per message, and the folded and squeezed readings are derived from the
    mapped normalized one rather than rebuilt.

    The mapped reading is compared with the reading the filter searches. They only differ if
    segment-wise normalization disagrees with whole-string normalization, in which case the maps fall
    back to whole whitespace-separated chunks and then to the whole message: a coarser map can hide
    more than the word, but never less.
    """
    if cache is None:
        cache = new_map_cache()

    cached = cache[kind]
    if cached is not None:
        return cached

    if kind == ReadingKind.NORMALIZED:
        mapped = _normalized(original)
    elif kind == ReadingKind.SQUEEZED:
        mapped = _derive(build(original, ReadingKind.NORMALIZED, cache), _squeeze_step)
    elif kind == ReadingKind.FOLDED:
        mapped = _derive(build(original, ReadingKind.NORMALIZED, cache), _fold_step)
    else:
        mapped = _derive(build(original, ReadingKind.FOLDED, cache), _squeeze_step)

    cache[kind] = mapped
    return mapped


def _normalized(original: str) -> MappedText:
    """The mapped normalized reading, checked against the reading the filter searched.

    Folding and squeezing are plain functions of their input, so once this one matches, every reading
    derived from it matches too.
    """
    map_: list[int] = []
    text = normalize_with_map(original, COMPARISON_STEPS, map_)
    expected = normalize_with_map(original, COMPARISON_STEPS, None)

    if text != expected:
        return chunk_map(original, expected)

    return MappedText(text, map_, map_)


def _derive(source: MappedText, step: _Step) -> MappedText:
    step_map: list[int] = []
    text = step(source.text, step_map)

    starts = [source.start_map[i] for i in step_map]
    if source.start_map is source.end_map:
        return MappedText(text, starts, starts)

    ends = [source.end_map[i] for i in step_map]
    return MappedText(text, starts, ends)


def chunk_map(original: str, reading: str) -> MappedText:
    """Map each whitespace-separated chunk of ``reading`` to the matching chunk of ``original``.

    Falls back to the whole message when the chunk counts differ.
    """
    original_chunks = _chunks(original)
    reading_chunks = _chunks(reading)
    if len(original_chunks) != len(reading_chunks):
        return whole_message_map(original, reading)

    starts = [0] * len(reading)
    ends = [0] * len(reading)
    chunk = 0
    start = 0
    end = 0

    for i in range(len(reading)):
        if chunk < len(reading_chunks) and i >= reading_chunks[chunk][0]:
            start = original_chunks[chunk][0]
            end = original_chunks[chunk][1]
            if i == reading_chunks[chunk][1]:
                chunk += 1

        # Separators keep the previous chunk's values, which keeps both maps non-decreasing.
        starts[i] = start
        ends[i] = end

    return MappedText(reading, starts, ends)


def whole_message_map(original: str, reading: str) -> MappedText:
    """Map every reading character to the whole message."""
    last = max(0, len(original) - 1)
    return MappedText(reading, [0] * len(reading), [last] * len(reading))


def _chunks(text: str) -> list[tuple[int, int]]:
    result: list[tuple[int, int]] = []
    start = -1

    for i in range(len(text) + 1):
        separator = i == len(text) or is_white_space(text[i])
        if not separator and start < 0:
            start = i
        elif separator and start >= 0:
            result.append((start, i - 1))
            start = -1

    return result
